### PR TITLE
Update `stdblib/shelve.pyi`

### LINKED ISSUE #5815: Make `shelve.Shelf` a generic type
It's more reasonable to write code as:

```python
import shelve
from shelve import Shelf

cache: Shelf[int] = shelve.open("~/.cache")
cache["apple"] = 1
```

than this:

```python
import shelve
from collections.abc import MutableMapping
from shelve import Shelve
from typing import cast

cache = cast(MutableMapping[str, int], shelve.open("~/.cache"))
cache["apple"] = 1
```

I suggest we change from:

```python
class Shelve(MutableMapping[Any, Any]):
```

to:

```python
class Shelve(MutableMapping[str, _VT]):
```

That being said, there could have been some nuances I didn't realize. Please don't hesitate to correct me.

### PR DESCRIPTION
**Pull request by MapleCCC**
Several things done:

1. Replace all occurrences of `Any` by respective concrete types.
2. Make `Shelf` and its subclassses generic. (fix #5815)
3. `shelve.open` should return a general `Shelf` object, not `DbfilenameShelf`. The documentation does not expose such implementation detail.
4. The argument `dict` is annotated with a covariant type `MutableMapping` rather than invariant type `Dict`.
5. Remove unnecessary methods. Some of them are inherited from `MutableMapping`, so no need to repeat them here.
6. Use builtin-in generics instead of importing from `typing`.

### CODE REVIEW

**Comment by MapleCCC** [comment]
Finally pass all tests. It's now ready for review.

**Comment by JelleZijlstra** [comment]
Haven't yet had a chance to look at all the code, but two quick comments:

> shelve.open should return a general Shelf object, not DbfilenameShelf. The documentation does not expose such implementation detail.

If it helps typing in practice, we're not averse to putting more precise types in the stubs where it makes sense. For example, for `builtins.open` we have overloads returning very specific undocumented IO classes. I'm not familiar with shelve, though, so I don't know what difference it makes here.

> The argument dict is annotated with a covariant type MutableMapping rather than invariant type Dict.

MutableMapping is also invariant in both of its type arguments. Mapping is covariant in its value argument but still invariant in its key argument. If possible, we should use Mapping instead.

**Comment by MapleCCC** [comment]
@JelleZijlstra Thanks for the kind reply and helpful information. I did make some mistakes in understanding the invariance/covariance stuff.

I am going to replace `MutableMapping` with `Mapping` per your suggestion.

As for the precise type returned by `shelve.open`, the only difference between `Shelf` and `DbfilenameShelf` viewing from outside is on their `__init__()` function, which is irrelevant in the case here. It should therefore be safe to use `Shelf` here. That being said, I am open to further opinions.

**Comment by MapleCCC** [comment]
On further introspection, the document and the implementation both mandates that the `dict` argument to the `Shelf.__init__()` call has to be `MutableMapping`. I will update then.

File: stdlib/shelve.pyi [hunk 1]
```diff
@@ -1,35 +1,39 @@
...
+from typing import Optional, Type, TypeVar, Union, overload
 
-class Shelf(collections.abc.MutableMapping[Any, Any]):
+_T = TypeVar("_T")
+_VT = TypeVar("_VT")
+
+class Shelf(MutableMapping[str, _VT]):
     def __init__(
-        self, dict: Dict[bytes, Any], protocol: Optional[int] = ..., writeback: bool = ..., keyencoding: str = ...
+        self, dict: MutableMapping[str, _VT], protocol: Optional[int] = ..., writeback: bool = ..., keyencoding: str = ...
```
Comment (JelleZijlstra) [inline stdlib/shelve.pyi:11]: The keys and values are both bytes. It always encodes keys before putting them in the underlying dict. The values are pickled representations of the logical values (you can see that in the `__getitem__` implementation).
```suggestion
        self, dict: MutableMapping[bytes, bytes], protocol: Optional[int] = ..., writeback: bool = ..., keyencoding: str = ...
```

File: stdlib/shelve.pyi [hunk 2]
```diff
@@ -1,35 +1,39 @@
...
     def __iter__(self) -> Iterator[str]: ...
     def __len__(self) -> int: ...
-    def __contains__(self, key: Any) -> bool: ...  # key should be str, but it would conflict with superclass's type signature
-    def get(self, key: str, default: Any = ...) -> Any: ...
-    def __getitem__(self, key: str) -> Any: ...
-    def __setitem__(self, key: str, value: Any) -> None: ...
+    @overload
+    def get(self, key: str) -> Optional[_VT]: ...
+    @overload
+    def get(self, key: str, default: _T = ...) -> Union[_VT, _T]: ...
```
Comment (JelleZijlstra) [inline stdlib/shelve.pyi:18]:
```suggestion
    def get(self, key: str, default: _T) -> Union[_VT, _T]: ...
```
The previous overload covers the case where default is not given.

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]

File: stdlib/shelve.pyi (hunk 1 above)
Comment (MapleCCC) [reply stdlib/shelve.pyi]: My bad. I did not read the implementation carefully enough. The fix is done.

File: stdlib/shelve.pyi (hunk 2 above)
Comment (MapleCCC) [reply stdlib/shelve.pyi]: Done.

**Comment by JelleZijlstra** [comment]
Thank you!

File: stdlib/shelve.pyi [hunk 3]
```diff
@@ -1,35 +1,39 @@
...
-from typing import Any, Dict, Iterator, Optional, Tuple
+from collections.abc import Iterator, MutableMapping
+from types import TracebackType
+from typing import Optional, Type, TypeVar, Union, overload
 
-class Shelf(collections.abc.MutableMapping[Any, Any]):
+_T = TypeVar("_T")
+_VT = TypeVar("_VT")
+
+class Shelf(MutableMapping[str, _VT]):
```
Comment (srittau) [inline stdlib/shelve.pyi:9]: Isn't `Generic` missing as base class? (Same for the classes below.)

Comment (JelleZijlstra) [reply stdlib/shelve.pyi:9]: We went over this recently :). You don't need Generic as a base class if you inherit from another generic class with a typevar in its type arguments.

Comment (srittau) [reply stdlib/shelve.pyi:9]: Considering how well my memory works, expect me to ask again next month. 😅

**Pull request merged**